### PR TITLE
fixes #14663 - support IPv6 DNS orchestration

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

File: app/models/concerns/orchestration/dns.rb [hunk 1]
```diff
@@ -13,163 +15,199 @@ def dns?
     (host.nil? || host.managed?) && managed? && hostname.present? && ip_available? && !domain.nil? && !domain.proxy.nil? && SETTINGS[:unattended]
   end
 
+  def dns6?
+    # host.managed? and managed? should always come first so that orchestration doesn't
+    # even get tested for such objects
+    (host.nil? || host.managed?) && managed? && hostname.present? && ip6_available? && !domain.nil? && !domain.proxy.nil? && SETTINGS[:unattended]
```
Comment (tbrisker) [inline app/models/concerns/orchestration/dns.rb]: Looks like
`SETTINGS[:unattended] && (host.nil? || host.managed?) && managed? && hostname.present?` is repeated quite a lot, how about extracting that to a separate method? (and checking the settings before all the other checks)

File: app/models/concerns/orchestration/dns.rb [hunk 2]
```diff
@@ -1,175 +1,211 @@
...
     begin
       a_record_result = recreate_a_record if dns?
+      aaaa_record_result = recreate_aaaa_record if dns6?
       ptr_record_result = recreate_ptr_record if reverse_dns?
-      a_record_result && ptr_record_result
+      ptr6_record_result = recreate_ptr6_record if reverse_dns6?
+      a_record_result && aaaa_record_result && ptr_record_result && ptr6_record_result
     rescue => e
-      Foreman::Logging.exception "Failed to rebuild DNS record for #{name}, #{ip}", e, :level => :error
+      Foreman::Logging.exception "Failed to rebuild DNS record for #{name}", e, :level => :error
```
Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb]: I think logging the IP of the interface (or ip6) is worth it here, so you know which interface is failing more easily

File: app/models/concerns/orchestration/dns.rb [hunk 3]
```diff
@@ -1,175 +1,211 @@
...
 
-  def del_dns_ptr_record_safe
-    if dns_ptr_record
-      begin
-        del_dns_ptr_record
-      rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DNS ptr_record for #{name}, #{ip}", e, :level => :error
+  RECORD_TYPES.each do |record_type|
+    define_method("del_dns_#{record_type}_record_safe") do # def del_dns_a_record_safe
+      if send(:"dns_#{record_type}_record") # dns_a_record
```
Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb]: These methods used to be protected, is there any reason to make them public now?

File: app/models/concerns/orchestration/dns.rb [hunk 4]
```diff
@@ -1,175 +1,211 @@
...
-    dns_a_record.conflicts.each { |c| c.destroy }
-  end
+    define_method("set_conflicting_dns_#{record_type}_record") do # def set_conflicting_dns_a_record
+      send(record).conflicts.each { |c| c.create } # dns_a_record.conflicts.each { |c| c.create }
+    end
 
-  def del_dns_ptr_record
-    dns_ptr_record.destroy
-  end
+    define_method("del_dns_#{record_type}_record") do # def del_dns_a_record
```
Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb]: I wonder if it's worth splitting this class now somehow.. gotta pick up [Refactoring](https://www.amazon.com/Refactoring-Improving-Design-Existing-Code/dp/0201485672) again...

Comment (timogoebel) [inline app/models/concerns/orchestration/dns.rb]: What do you have in mind @dLobatog ?

Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb]: Nothing so far, I'm running codeclimate on this & trying to read more about how to fix this kind of duplication in a neat and easy to understand way

Comment (timogoebel) [inline app/models/concerns/orchestration/dns.rb]: I thought about that for a wile as well, but did not come up with a neat way to fix this. Just copying the code and living with the duplication is no option imho. I made mistakes all over. The define method thing is also not very pretty and readable. However we do have something like this in the codebase already. Please let me know if you have any suggestions!

File: app/models/concerns/orchestration/dns.rb [hunk 5]
```diff
@@ -2,179 +2,224 @@ module Orchestration::DNS
...
-  def dns?
+  def dns_ready?
     # host.managed? and managed? should always come first so that orchestration doesn't
     # even get tested for such objects
-    (host.nil? || host.managed?) && managed? && hostname.present? && ip_available? && !domain.nil? && !domain.proxy.nil? && SETTINGS[:unattended]
+    SETTINGS[:unattended] && (host.nil? || host.managed?) && managed? && hostname.present?
+  end
+
+  def dns?
+    dns_ready? && ip_available? && !domain.nil? && !domain.proxy.nil?
```
Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb]: nitpick, `variable.present?` seems more readable than `!variable.nil?` - applies to many lines around this one

File: app/models/concerns/orchestration/dns.rb [hunk 6]
```diff
@@ -2,179 +2,224 @@ module Orchestration::DNS
...
 
   def dns_a_record
     return unless dns? or @dns_a_record
     handle_validation_errors do
-      @dns_a_record ||= Net::DNS::ARecord.new dns_record_attrs
+      @dns_a_record ||= Net::DNS::ARecord.new dns_a_record_attrs
+    end
+  end
+
+  def dns_aaaa_record
```
Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb]: It seems like the creation/deletion parts of this class have outgrown the responsibility of this class (just maintaining the validations/callbacks to create/delete DNS records during orchestration. I think it'd be best to move creation and deletion to their own modules like:

``` ruby
module DNSRecordCreation
  def create_dns_record(type)
    return if instance_variable_get("@dns_#{type}_record").present?
    case type
    when 'aaaa', 'ptr6'
      return unless dns6? && ip6_available?
    when 'a', 'ptr' 
      return unless dns? && ip_available?
    end

    handle_validation_errors do
      instance_variable_set(
        "@dns_#{type}_record", 
        dns_class(type).new(send(:"dns_#{type}_record_attrs"))
    end
  end

  private

  def dns_class(type)
    class_name = type.upcase.constantize
    Net::DNS::class_name
  end

  def dns_a_record_attrs
     { :hostname => hostname, :ip => ip, :resolver => domain.resolver, :proxy => domain.proxy }
   end

  def dns_aaaa_record_attrs
    { :hostname => hostname, :ip => ip6, :resolver => domain.resolver, :proxy => domain.proxy }
  end

   def reverse_dns_record_attrs
     { :hostname => hostname, :ip => ip, :proxy => subnet.dns_proxy }
   end

  def reverse_dns6_record_attrs
    { :hostname => hostname, :ip => ip6, :proxy => subnet6.dns_proxy }
  end
end
```

Creation and deletion of DNS entries would go like `create_dns_record('aaaa')` and it seems like at least each file would have its own responsibility.

**Comment by dLobatog** [comment]
@timogoebel Thanks for the updates, I'll test creating a IPv6 subnet manually and assigning it to a host then creating it. I think the creation/deletion of hosts can be moved to a module of their own as per https://github.com/theforeman/foreman/pull/3567#discussion_r66823322 , after that testing it should be easy.

**Comment by timogoebel** [comment]
@dLobatog : Thanks for your comments. I followed your suggestions. Had to change queue processing a little for this to work. Also noticed I forgot to move some tests when I split the large PR. Happy to hear your comments. Disclaimer: Haven't tested this well, so far.

File: app/models/concerns/dns_interface.rb [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+  def get_dns_record(type)
+    instance_variable_get("@dns_#{type}_record")
+  end
+
+  def store_dns_record(type, value)
+    instance_variable_set("@dns_#{type}_record", value)
+  end
+
+  def validate_record_type!(type)
```
Comment (dLobatog) [inline app/models/concerns/dns_interface.rb]: Nitpick, `!` is used in Ruby for methods that change state (either the object passed or global state), this one doesn't change state, it'd just raise an exception.

File: app/models/concerns/dns_interface.rb [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    raise ArgumentError, "#{type} is not a valid dns record type" unless RECORD_TYPES.include?(type)
```
Comment (dLobatog) [inline app/models/concerns/dns_interface.rb]: `::Foreman::Exception.new(N_("%s is not a valid DNS record type"), type)`

File: app/models/subnet.rb [hunk 9]
```diff
@@ -23,7 +23,9 @@ class Subnet < ActiveRecord::Base
 
   # This casts Subnet to Subnet::Ipv4 if no type is set
   def self.new(*attributes, &block)
-    return Subnet::Ipv4.new_without_cast(*attributes, &block) if self == Subnet
+    type = nil
```
Comment (dLobatog) [inline app/models/subnet.rb]: I don't think you need to initialize `type` to nil, it's already nil by default isn't it?

File: app/models/concerns/orchestration/dns.rb [hunk 10]
```diff
@@ -8,173 +8,108 @@ module Orchestration::DNS
...
 
-  def del_dns_ptr_record_safe
-    if dns_ptr_record
       begin
-        del_dns_ptr_record
+        results[record_type] = recreate_dns_record(record_type) if dns_feasible?(record_type)
       rescue => e
-        Foreman::Logging.exception "Proxy failed to delete DNS ptr_record for #{name}, #{ip}", e, :level => :error
+        Foreman::Logging.exception "Failed to rebuild DNS record for #{name}(#{ip}/#{ip6})", e, :level => :error
+        return false
```
Comment (dLobatog) [inline app/models/concerns/orchestration/dns.rb:51]: nitpick, no need for explicit return here

**Comment by dLobatog** [comment]
@timogoebel Awesome job! I left a few minor comments, I'm going to test this a bit with Libvirt and see how far I get.

File: app/models/subnet.rb (hunk 9 above)
Comment (timogoebel) [inline app/models/subnet.rb]: I think, we need it when `(h = attributes.first).is_a?(Hash)` is false, don't we?

File: app/models/concerns/orchestration/dns.rb (hunk 10 above)
Comment (timogoebel) [inline app/models/concerns/orchestration/dns.rb:51]: We do need it. The diff is a little misleading. `results.values.all?` comes after this block.

**Comment by timogoebel** [comment]
@dLobatog : Fixed the nitpicks. Thanks 👍

File: app/models/subnet.rb (hunk 9 above)
Comment (dLobatog) [inline app/models/subnet.rb]: Nope, `type` is a db field of this model so it's nil by default. Try removing line 26, then putting a debugger breakpoint and you'll see `type` is nil.

**Comment by dLobatog** [comment]
Worked well when creating/deleting a host with DNS & warned about conflict. I used the UI PR to create the subnet and interface https://github.com/theforeman/foreman/pull/3587/files

**Comment by dLobatog** [comment]
Merged as 041436c8e2eb9528a7f7c1c760621d1639100533, thanks @timogoebel!

**Pull request closed without merging**